### PR TITLE
fix: do not migrate during connection close

### LINKED ISSUE #2569: crash during connection migration
```
F20240211 10:58:17.426013 1814968 uring_socket.cc:373] Check failed: !error_cb_wrapper_ 
*** Check failure stack trace: ***
    @     0x55cd16b6ada7  google::LogMessage::SendToLog()
    @     0x55cd16b632e3  google::LogMessage::Flush()
    @     0x55cd16b64c7f  google::LogMessageFatal::~LogMessageFatal()
    @     0x55cd168c6587  util::fb2::UringSocket::RegisterOnErrorCb()
    @     0x55cd16481608  facade::Connection::OnPostMigrateThread()
    @     0x55cd16894b75  util::ListenerInterface::Migrate()
    @     0x55cd1647e1cb  facade::Connection::HandleMigrateRequest()
    @     0x55cd16481b04  facade::Connection::IoLoop()
    @     0x55cd16481eac  facade::Connection::ConnectionFlow()
    @     0x55cd16482e0d  facade::Connection::HandleRequests()
    @     0x55cd16894233  util::ListenerInterface::RunSingleConnection()

```

to reproduce, run dragonfly in build-opt: `./dragonfly --logtostderr --proactor_threads=2`
then  load the lua script with `redis-cli --eval ./fill.lua k2 , 100`
```lua
local function randomString(length)
	local chars = {}
	for i = 1, length do
		chars[i] = string.char(math.random(33, 126))
	end
	return table.concat(chars)
end

local upper = tonumber(ARGV[1])

for i = 1, upper do
    redis.call('zadd', KEYS[1], i, randomString(10))
end
```

and then run memtier: `memtier_benchmark  --command "evalsha f3be7dc8bd0e838ee3920321ce6999e9ae888e0f 1 __key__ 10000" --hide-histogram -n 1000 --distinct-client-seed --threads=2 -c 2 --key-maximum=1000000`

### PR DESCRIPTION
**Pull request by romange**
Fixes #2569
Before the change we had a corner case where Dragonfly would call OnPreMigrateThread but would not call CancelOnErrorCb because OnBreakCb has already been called (it resets break_cb_engaged_)

On the other hand in OnPostMigrateThread we called RegisterOnErrorCb if breaker_cb_ which resulted in double registration. This change simplifies the logic by removing break_cb_engaged_ flag since CancelOnErrorCb is safe to call if nothing is registered. Moreover, we now skip Migrate flow if a socket is being closed.

### CODE REVIEW

**Review by kostasrim — DISMISSED** [review]

**Review by chakaz — APPROVED** [review]

**Comment by romange** [comment]
@chakaz  with help of @adiholden  I added the test that sporadically reproduced the problem for me

**Review by adiholden — APPROVED** [review]

File: tests/dragonfly/eval_test.py [hunk 1]
```diff
@@ -310,3 +314,24 @@ async def measure_blocked():
...
+
+    reader, writer = await asyncio.open_connection("localhost", df_server.port)
+
+    # write a EVALSHA that will ask for migration (75% it's on the wrong shard)
+    writer.write((f"EVALSHA {sha} 1 a\r\n").encode())
+    await writer.drain()
+
+    # disconnect the client connection
+    writer.close()
+    await writer.wait_closed()
```
Comment (chakaz) [inline tests/dragonfly/eval_test.py:337]: Maybe do this in a loop many times to increase chances of catching this and other bugs?

**Review by chakaz — DISMISSED** [review]

File: tests/dragonfly/eval_test.py (hunk 1 above)
Comment (romange) [reply tests/dragonfly/eval_test.py:337]: when trying to got with this suggestion, I discovered that our pytest framework does not recognise when dragonfly process crashes 😮 . sending a fix for this as well.

**Review by chakaz — APPROVED** [review]
nice :)

**Review by chakaz — APPROVED** [review]

**Pull request merged**